cmd/crypto/vault.go

```diff
@@ -130,79 +130,68 @@ func NewVault(config VaultConfig) (KMS, error) {
 	return v, nil
 }
 
-// renewSecret tries to renew the given secret. It blocks
-// until it receives either the new secret or encounters an error.
-func (v *vaultService) renewSecret(secret *vault.Secret) (*vault.Secret, error) {
+// reauthenticate() tries to login in 1 minute
+// intervals until successful.
+func (v *vaultService) reauthenticate() {
+	retryDelay := 1 * time.Minute
+	go func() {
+		for {
+			if err := v.authenticate(); err != nil {
+				time.Sleep(retryDelay)
+				continue
+			}
+			return
+		}
+	}()
+}
+
+// renewer calls vault client's renewer that automatically
+// renews secret periodically
+func (v *vaultService) renewer(secret *vault.Secret) {
 	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
 		Secret: secret,
 	})
 	if err != nil {
 		logger.CriticalIf(context.Background(), fmt.Errorf("crypto: failed to create hashicorp vault renewer: %s", err))
 	}
+	v.tokenRenewer = renewer
 	go renewer.Renew()
 	defer renewer.Stop()
 
 	for {
 		select {
 		case err := <-renewer.DoneCh():
 			if err != nil {
-				return nil, err
+				v.reauthenticate()
+				return
 			}
-		case renew := <-renewer.RenewCh():
-			if renew.Secret == nil || renew.Secret.Auth == nil {
-				return nil, ErrKMSAuthLogin
-			}
-			return renew.Secret, nil
+
+			// Renewal is now over
+		case <-renewer.RenewCh():
+			return
```

Comment:
We shouldn't return here @poornas we need renewer to be running always .. i.e RenewCh() will return the renewed token. We also don't need a new `v.tokenRenewer` its not used by anyone..

Reply:
The renewed token returned by RenewCh should be used again to set the AuthToken() .and also used with the NewRenewer()

Reply:
when we reauthenticate, we start a new renewer

Reply:
We shouldn't - that is not the flow from vault SDK we are incurring costs which are not needed.. @poornas

Reply:
https://godoc.org/github.com/hashicorp/vault/api#Renewer

Reply:
Also notice that the renewed secret may be `nil` - See: https://github.com/minio/minio/pull/7259/files#diff-bc9558a54da677c8f7933fee0f0c42f5L152